### DISCUSSION TITLE
Errors - user agent required, invalid api key & how to preload ticker data

### DISCUSSION THREAD

**Discussion by MrAdamo** [mr-highball/simplebot-support#12] [Q&A] [answered]
I saw this error while checking the log and have no idea what it is about.

-ERROR- 13/01/2021 23:01:43 - BTC-EUR - invalid web status code 400
{"message":"User-Agent header is required."}

Was the bot attempting to trade and failed?

**Comment by mr-highball** [comment, marked as answer]
@MrAdamo if you uncheck "info" & "warning" but check "error" for the logging and have the bot run, does this error message keep repeating? the user-agent is just a piece of information passed in each web call that identifies "what" is communicating with the other server (in this case coinbase server)
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent
![image](https://user-images.githubusercontent.com/35942565/104518055-168a1b00-55c5-11eb-838a-ddc73a348682.png)

in addition to my previous question about logging can you answer below:
are you running in sandbox / prod?
do you have Key/Secret/Passphrase filled out correctly on the authenticator tab?
can you verify that no orders were placed on coinbase?

**Reply by mr-highball** [reply]
for a little more backstory, the user-agent _should_ always be sending for each request since I've built this into the coinbase pro api to mimic a "firefox" browser. 

if we can't figure out with the few questions above, I do have a tester application included with the coinbase library that I've attached below.
this little utility is bundled with the source and _could_ be built from scratch, but for ease I've attached a zip of a prebuilt version.
![image](https://user-images.githubusercontent.com/35942565/104518585-e3945700-55c5-11eb-9f81-b9c8d73af48b.png)
[coinbase_tester_01_13_2021.zip](https://github.com/mr-highball/simplebot-support/files/5811399/coinbase_tester_01_13_2021.zip)

basically the screenshot shows the utility after the _test_ button was clicked and that indeed I have a user agent. additionally if credentials were put in (like the authenticator tab on simple-bot) then the "base web client" tab can be clicked to see if a valid connection can be made. if you use sandbox creds, then the "order" tab can be used to see if an order can be placed 
![image](https://user-images.githubusercontent.com/35942565/104519132-f0657a80-55c6-11eb-8b1e-fe7306d1c4b1.png)

again, let's try to figure this out with simple questions first, but putting the "next steps" if we have to

**Reply by MrAdamo** [reply]
I run the bot with Errors and Warnings always checked I manually check the Info from time to time.
The bot has been running for hours and I saw this error only once.
It runs live with all the API codes in place, so far it hasn't placed any order.
It's pretty late now so I don't have time to use the tester now.

**Reply by mr-highball** [reply]
If it's only been a few hours (and you didn't preload tickers) it's normal not to have any orders placed for a while. no orders will be placed until the strategies are "ready" (collected enough tickers to satisfy their data needs and is dependent on your timeframe setting). If you see the problem persist just report back here and we can look back into it. I'll leave the discussion open and if I don't hear back in a while I can assume it got straightened out.

**Reply by MrAdamo** [reply]
The bot has been running for a couple of days. I preloaded a day worth of log but I noticed that you can only load one file while the bot create a new file every day, should I merge all the files together? Will a simply copy and paste in the notepad do the trick?
The error happened only once so I was worried it was related to a buy order, silly me.
Also should I always leave the "info" checked in the strategy logger? If I don't and close the bot, will it lose any data useful for said strategy?

**Reply by mr-highball** [reply]
you can "multi select" ticker files to load either by holding CTRL + clicking or by doing a drag selection
![image](https://user-images.githubusercontent.com/35942565/104628385-29a0f780-5666-11eb-92c4-736627ebdf3b.png)
so no need for you to merge them your self. so when you stop the bot for whatever reason you can just select as many recent files as your strategy needs (probably the last 3 or 4 files).

A side note to that, the simulator actually can do a decimate & merge & save to a single file for you (in the case you wanted to simulate with newer data and skip having to decimate)
![image](https://user-images.githubusercontent.com/35942565/104628664-7c7aaf00-5666-11eb-85e5-ee3a89d445e9.png)

you also do not need to keep info checked, it's just there to get a full picture of what's going on, but definitely not necessary

**Reply by MrAdamo** [reply]
The same error appeared again:
-ERROR- 15/01/2021 13:53:44 - BTC-EUR - invalid web status code 400
{"message":"User-Agent header is required."}
-ERROR- 15/01/2021 14:31:58 - BTC-EUR - invalid web status code 400
{"message":"User-Agent header is required."}

And also other errors on other instances of the bot I have running. Each bot has it's own separate folder.
-ERROR- 15/01/2021 13:31:15 - ETH-EUR - invalid web status code -3
The bot above has been collecting ETH ticker files. With 0 fund and a valid API key.

-ERROR- 15/01/2021 14:11:00 - TGDAXOrderManagerImpl::Place::an error occurrred [error]:invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:11:00 - TGDAXBobberStrategyImpl::OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:11:00 - OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:11:01 - TGDAXOrderManagerImpl::Place::an error occurrred [error]:invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:11:01 - TGDAXBobberStrategyImpl::OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:11:01 - OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:12:00 - TGDAXOrderManagerImpl::Place::an error occurrred [error]:invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:12:00 - TGDAXBobberStrategyImpl::OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:12:00 - OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:12:01 - TGDAXOrderManagerImpl::Place::an error occurrred [error]:invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:12:01 - TGDAXBobberStrategyImpl::OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:12:01 - OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:14:28 - TGDAXOrderManagerImpl::Place::an error occurrred [error]:invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:14:28 - TGDAXBobberStrategyImpl::OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:14:28 - OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:14:30 - TGDAXOrderManagerImpl::Place::an error occurrred [error]:invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:14:30 - TGDAXBobberStrategyImpl::OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
-ERROR- 15/01/2021 14:14:30 - OpenPosition::invalid web status code 401
{"message":"Invalid API Key"}
While this bot has been running in sandbox mode with sandbox put as Key/Secret/Passphrase in all the API key setting like you explained in your Software Overview Wiki.

**Reply by mr-highball** [reply]
so the "invalid api key" message is coming straight from coinbase, so I'm thinking one of the three (key/secret/passphrase) is off OR that the api key does not have the "TRADE" permission. Since it's a sandbox key it also has to be made on the sandbox url
* https://public.sandbox.pro.coinbase.com/ (login here and make sure funds are there and create the key here)

AND you have to "fund" the sandbox account (same process has funding regular account with the deposit button, but instead of linking a bank account it just lets you type in how much money you have).
I'd recommend just blowing away this key and just re-creating following their steps (make sure you copy the secret once it's shown because it's only shown once, and give it TRADE permission)
* https://help.coinbase.com/en/pro/other-topics/api/how-do-i-create-an-api-key-for-coinbase-pro
* https://docs.pro.coinbase.com/#generating-an-api-key

I'm still uncertain on the user agent error since I have not received that message before with valid creds, but the tester application I linked in the previous comment can be used to do the preliminary checking that your key is good and that it can place a sandbox order.

Lastly for the "invalid web status code -3" if it doesn't happen too often this may be just fine and a problem with communication to their api (sometimes they have outages or timeouts occur). I'd recommend if your running multiple instances (which is no problem I do it too) to offset the poll interval by a second or so (if one bot is set to 1500 set the second one to 2500 or 3000 msecs). This is to avoid any issues hitting the ticker endpoint since coinbase does "throttle" the amount of requests per second
https://docs.pro.coinbase.com/#rate-limits

down the road I'll probably give the option to connect to their websocket feed which doesn't have the rate limits, but I just haven't gotten to it.

**Reply by mr-highball** [reply]
one more thing I should mention, just make sure the "sandbox" mode is checked on the authenticator tab of the bot
![image](https://user-images.githubusercontent.com/35942565/104752227-69c6af80-5724-11eb-9ca5-2c791e99e57c.png)

**Reply by mr-highball** [reply]
have not heard back, hope you were able to resolve the issues. if not I'd recommend trying the latest v1.0.1 release.
